In [37]:
import pandas as pd

df_ind = pd.read_csv('database_IND.csv')
df_aus = pd.read_csv('database_AUS.csv')
df_usa = pd.read_csv('database_USA.csv')

In [38]:
data = pd.concat([df_ind , df_aus , df_usa])

In [39]:
data.head()

country country_long                      name   gppd_idnr  capacity_mw  \
0     IND        India          ACME Solar Tower  WRI1020239          2.5   
1     IND        India       ADITYA CEMENT WORKS  WRI1019881         98.0   
2     IND        India  AES Saurashtra Windfarms  WRI1026669         39.2   
3     IND        India               AGARTALA GT  IND0000001        135.0   
4     IND        India              AKALTARA TPP  IND0000002       1800.0   

   latitude  longitude primary_fuel other_fuel1 other_fuel2  ...  \
0   28.1839    73.2407        Solar         NaN         NaN  ...   
1   24.7663    74.6090         Coal         NaN         NaN  ...   
2   21.9038    69.3732         Wind         NaN         NaN  ...   
3   23.8712    91.3602          Gas         NaN         NaN  ...   
4   21.9603    82.4091         Coal         Oil         NaN  ...   

  year_of_capacity_data  generation_gwh_2013 generation_gwh_2014  \
0                   NaN                  NaN                 NaN   
1                   NaN                  NaN                 NaN   
2                   NaN                  NaN                 NaN   
3                2019.0              2.41461          617.789264   
4                2019.0                  NaN         3035.550000   

  generation_gwh_2015 generation_gwh_2016 generation_gwh_2017  \
0                 NaN                 NaN                 NaN   
1                 NaN                 NaN                 NaN   
2                 NaN                 NaN                 NaN   
3             843.747          886.004428          663.774500   
4            5916.370         6243.000000         5385.579736   

   generation_gwh_2018  generation_gwh_2019         generation_data_source  \
0                  NaN                  NaN                            NaN   
1                  NaN                  NaN                            NaN   
2                  NaN                  NaN                            NaN   
3           626.239128                  NaN  Central Electricity Authority   
4          7279.000000                  NaN  Central Electricity Authority   

   estimated_generation_gwh  
0                       NaN  
1                       NaN  
2                       NaN  
3                       NaN  
4                       NaN  

[5 rows x 27 columns]

In [40]:
data.tail()

country              country_long                           name  \
11874     USA  United States of America       Zotos International WPGF   
11875     USA  United States of America  Zumbro Community Solar Garden   
11876     USA  United States of America            Zumbro Solar Garden   
11877     USA  United States of America                           Zuni   
11878     USA  United States of America            eBay - South Jordan   

        gppd_idnr  capacity_mw  latitude  longitude primary_fuel other_fuel1  \
11874  USA0057648          3.4   42.8869   -76.9683         Wind         NaN   
11875  USA0061574          1.0   44.3195   -92.6703        Solar         NaN   
11876  USA0063426          0.0   44.0460   -92.7600          NaN         NaN   
11877  USA0000478          0.0   39.7373  -105.0170          NaN         NaN   
11878  USA0059281          9.8   40.5606  -112.0478          Gas         NaN   

      other_fuel2  ... year_of_capacity_data  generation_gwh_2013  \
11874         NaN  ...                2019.0                3.671   
11875         NaN  ...                2019.0                  NaN   
11876         NaN  ...                2019.0                  NaN   
11877         NaN  ...                2019.0                0.000   
11878         NaN  ...                2019.0               12.968   

      generation_gwh_2014 generation_gwh_2015 generation_gwh_2016  \
11874             2.46015               2.489               1.670   
11875                 NaN                 NaN                 NaN   
11876                 NaN                 NaN                 NaN   
11877             0.00000               0.000                 NaN   
11878            51.27600              71.505              76.392   

      generation_gwh_2017  generation_gwh_2018  generation_gwh_2019  \
11874             1.23349                2.630                1.211   
11875                 NaN                1.161                1.589   
11876                 NaN                  NaN                  NaN   
11877                 NaN                  NaN                  NaN   
11878            76.63100               75.228             72.06506   

                       generation_data_source  estimated_generation_gwh  
11874  U.S. Energy Information Administration                       NaN  
11875  U.S. Energy Information Administration                       NaN  
11876                                     NaN                       NaN  
11877  U.S. Energy Information Administration                       NaN  
11878  U.S. Energy Information Administration                       NaN  

[5 rows x 27 columns]

In [41]:
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt


In [42]:
data.shape

(13298, 27)

*We have 13298 rows and 27 columns in our final dataset.*

In [10]:
# Checking for null values

data.isnull().sum()

country                        0
country_long                   0
name                           0
gppd_idnr                      0
capacity_mw                    6
latitude                      63
longitude                     63
primary_fuel                1091
other_fuel1                 7022
other_fuel2                 7878
other_fuel3                 7984
commissioning_year          1696
owner                        587
source                         0
url                            0
geolocation_source            21
wepp_id                     8036
year_of_capacity_data        416
generation_gwh_2013         3704
generation_gwh_2014         3510
generation_gwh_2015         3235
generation_gwh_2016         2842
generation_gwh_2017         2506
generation_gwh_2018         2170
generation_gwh_2019         2542
generation_data_source      1615
estimated_generation_gwh    8036
dtype: int64

*Remove those columns , which has maximum number of null values because If we fill those values than prediction will not correct.*

In [43]:
data.drop(['estimated_generation_gwh' , 'wepp_id' , 'other_fuel3' , 'other_fuel2' , 'other_fuel1'] , axis = 1 , inplace = True)

In [44]:
data.isnull().sum()

country                      0
country_long                 0
name                         0
gppd_idnr                    0
capacity_mw                  6
latitude                    78
longitude                   78
primary_fuel              2004
commissioning_year        2644
owner                      593
source                       0
url                          0
geolocation_source          23
year_of_capacity_data      438
generation_gwh_2013       6128
generation_gwh_2014       5764
generation_gwh_2015       5305
generation_gwh_2016       4597
generation_gwh_2017       4006
generation_gwh_2018       3440
generation_gwh_2019       3520
generation_data_source    2463
dtype: int64

*Removing unnecessarry columns*

In [13]:
data.drop(['gppd_idnr','owner','generation_data_source'],axis=1,inplace=True)

In [45]:
data.describe()  #Checking for statistical summary of DataSet


capacity_mw      latitude     longitude  commissioning_year  \
count  13292.000000  13220.000000  13220.000000        10654.000000   
mean     117.738717     35.053213    -73.794217         1998.184904   
std      347.169038     15.444195     62.868834           24.017781   
min        0.000000    -42.888300   -171.712400         1896.000000   
25%        1.500000     33.901700   -104.659875         1988.000000   
50%        5.500000     38.389750    -87.346000         2008.000000   
75%       52.725000     42.175125    -75.475925         2015.000000   
max     6809.000000    425.246100    153.434100         2020.000000   

       year_of_capacity_data  generation_gwh_2013  generation_gwh_2014  \
count                12860.0          7170.000000          7534.000000   
mean                  2019.0           563.383222           635.739731   
std                      0.0          2030.523760          2193.172655   
min                   2019.0          -947.600000          -989.619000   
25%                   2019.0             1.653075             2.018000   
50%                   2019.0            21.384500            22.241000   
75%                   2019.0           185.326498           217.069500   
max                   2019.0         31431.080000         32320.917000   

       generation_gwh_2015  generation_gwh_2016  generation_gwh_2017  \
count          7993.000000          8701.000000          9292.000000   
mean            616.801877           561.784508           532.472482   
std            2148.167470          2050.411295          2001.411908   
min            -864.428000          -768.620000          -934.944000   
25%               2.140000             2.108000             2.063750   
50%              19.504000            15.410278            13.433861   
75%             200.279000           173.348004           157.889688   
max           32525.595000         32377.477000         35116.000000   

       generation_gwh_2018  
count          9858.000000  
mean            513.679016  
std            1963.485305  
min            -982.622000  
25%               2.141250  
50%              12.308500  
75%             149.141249  
max           35136.000000

In [46]:
data.nunique()

country                       3
country_long                  3
name                      13266
gppd_idnr                 13298
capacity_mw                2304
latitude                  12385
longitude                 12733
primary_fuel                 14
commissioning_year         1936
owner                      5877
source                      206
url                         325
geolocation_source            5
year_of_capacity_data         1
generation_gwh_2013        5616
generation_gwh_2014        6306
generation_gwh_2015        6692
generation_gwh_2016        7150
generation_gwh_2017        7630
generation_gwh_2018        8040
generation_gwh_2019        8368
generation_data_source        3
dtype: int64

*Removing those columns, which has very less number of unique values and also those columns will be removed which has no use*

In [47]:
data.drop(['year_of_capacity_data' , 'name' , 'url'] ,axis = 1 , inplace = True)

In [48]:
data['geolocation_source'].unique()

array(['National Renewable Energy Laboratory', 'WRI', nan,
       'Industry About',
       'Australian Renewable Energy Mapping Infrastructure',
       'U.S. Energy Information Administration'], dtype=object)

*Here we can see that there is NAN present in column, So we need to fill this*

In [49]:
# filling geolocation_source with mode because category data

data['geolocation_source'].fillna(data['geolocation_source'].mode()[0] , inplace = True)

In [50]:
data['geolocation_source'].isnull().sum()

0

In [51]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13298 entries, 0 to 11878
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   country                 13298 non-null  object 
 1   country_long            13298 non-null  object 
 2   gppd_idnr               13298 non-null  object 
 3   capacity_mw             13292 non-null  float64
 4   latitude                13220 non-null  float64
 5   longitude               13220 non-null  float64
 6   primary_fuel            11294 non-null  object 
 7   commissioning_year      10654 non-null  float64
 8   owner                   12705 non-null  object 
 9   source                  13298 non-null  object 
 10  geolocation_source      13298 non-null  object 
 11  generation_gwh_2013     7170 non-null   float64
 12  generation_gwh_2014     7534 non-null   float64
 13  generation_gwh_2015     7993 non-null   float64
 14  generation_gwh_2016     8701 non-null   flo

In [52]:
# filling all the null values with median in numrical values

# # Calculate the median for each column
# medians = data[['capacity_mw', 'latitude', 'longitude', 'commissioning_year',
#                 'generation_gwh_2013', 'generation_gwh_2014', 'generation_gwh_2015',
#                 'generation_gwh_2016', 'generation_gwh_2017', 'generation_gwh_2018',
#                 'generation_gwh_2019']].median()

# # Fill missing values in each column with its median
# data.fillna(medians, inplace=True)

# import pandas as pd

# # Remove rows with non-numerical values in the 'generation_gwh_2019' column
# data = data[pd.to_numeric(data['generation_gwh_2019'], errors='coerce').notnull()]

# # Calculate the median for each column
# medians = data[['capacity_mw', 'latitude', 'longitude', 'commissioning_year',
#                 'generation_gwh_2013', 'generation_gwh_2014', 'generation_gwh_2015',
#                 'generation_gwh_2016', 'generation_gwh_2017', 'generation_gwh_2018',
#                 'generation_gwh_2019']].median()

# # Fill missing values in each column with its median
# data.fillna(medians, inplace=True)



# Calculate the mode for each specified column
modes = data[['capacity_mw', 'latitude', 'longitude', 'commissioning_year',
              'generation_gwh_2013', 'generation_gwh_2014', 'generation_gwh_2015',
              'generation_gwh_2016', 'generation_gwh_2017', 'generation_gwh_2018',
              'generation_gwh_2019']].mode().iloc[0]

# Fill missing values in each column with its mode
data.fillna(modes, inplace=True)


In [53]:
data.isnull().sum()

country                      0
country_long                 0
gppd_idnr                    0
capacity_mw                  0
latitude                     0
longitude                    0
primary_fuel              2004
commissioning_year           0
owner                      593
source                       0
geolocation_source           0
generation_gwh_2013          0
generation_gwh_2014          0
generation_gwh_2015          0
generation_gwh_2016          0
generation_gwh_2017          0
generation_gwh_2018          0
generation_gwh_2019          0
generation_data_source    2463
dtype: int64

In [24]:
import pandas as pd

# Remove rows with non-numerical values in the 'primary_fuel' column
data = data[pd.to_numeric(data['primary_fuel'], errors='coerce').notnull()]

In [25]:
# Calculate the median of the 'primary_fuel' column
med = data['primary_fuel'].median()


In [26]:
# Fill missing values in the 'primary_fuel' column with the median
data.fillna(med, inplace=True)

In [27]:
data.isnull().sum()

country                0.0
country_long           0.0
capacity_mw            0.0
latitude               0.0
longitude              0.0
primary_fuel           0.0
commissioning_year     0.0
source                 0.0
geolocation_source     0.0
generation_gwh_2013    0.0
generation_gwh_2014    0.0
generation_gwh_2015    0.0
generation_gwh_2016    0.0
generation_gwh_2017    0.0
generation_gwh_2018    0.0
generation_gwh_2019    0.0
dtype: float64

In [28]:
data.nunique()

country                0
country_long           0
capacity_mw            0
latitude               0
longitude              0
primary_fuel           0
commissioning_year     0
source                 0
geolocation_source     0
generation_gwh_2013    0
generation_gwh_2014    0
generation_gwh_2015    0
generation_gwh_2016    0
generation_gwh_2017    0
generation_gwh_2018    0
generation_gwh_2019    0
dtype: int64

In [29]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 0 entries
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   country              0 non-null      object 
 1   country_long         0 non-null      object 
 2   capacity_mw          0 non-null      float64
 3   latitude             0 non-null      float64
 4   longitude            0 non-null      float64
 5   primary_fuel         0 non-null      object 
 6   commissioning_year   0 non-null      float64
 7   source               0 non-null      object 
 8   geolocation_source   0 non-null      object 
 9   generation_gwh_2013  0 non-null      float64
 10  generation_gwh_2014  0 non-null      float64
 11  generation_gwh_2015  0 non-null      float64
 12  generation_gwh_2016  0 non-null      float64
 13  generation_gwh_2017  0 non-null      float64
 14  generation_gwh_2018  0 non-null      float64
 15  generation_gwh_2019  0 non-null      float64
dtypes: floa

In [32]:
# sns.set_style("whitegrid")
# sns.countplot(x="primary_fuel", data=data)

# **EDA : EXPLORATORY DATA ANALYSIS**

In [34]:
data.head(10)

Empty DataFrame
Columns: [country, country_long, capacity_mw, latitude, longitude, primary_fuel, commissioning_year, source, geolocation_source, generation_gwh_2013, generation_gwh_2014, generation_gwh_2015, generation_gwh_2016, generation_gwh_2017, generation_gwh_2018, generation_gwh_2019]
Index: []

In [35]:
data.corr()

(0, 16)

In [36]:
sns.heatmap(data.corr , annot = True , linewidth=1)

Empty DataFrame
Columns: [country, country_long, capacity_mw, latitude, longitude, primary_fuel, commissioning_year, source, geolocation_source, generation_gwh_2013, generation_gwh_2014, generation_gwh_2015, generation_gwh_2016, generation_gwh_2017, generation_gwh_2018, generation_gwh_2019]
Index: []


*Here we can notice generation_gwh_2013, generation_gwh_2014, generation_gwh_2015, generation_gwh_2016, generation_gwh_2017 are positively correlated to capacity_mw and latitude is negative correlated*

In [ ]:
plt.figure(figsize = (8 , 4))
sns.scatterplot(x = 'capacity_mw' , y = 'generation_gwh_2013' , data = data)

*-> We can notice here as electricity generation in gigawatt-hour increasing also increasing the no of electrical generating capacity in megawatts in 2013*

In [ ]:
plt.figure(figsize=(8,4))
sns.scatterplot(x='capacity_mw',y='generation_gwh_2014',data=df)

*In 2014 we can also notice here as electricity generation in gigawatt-hour increasing also increasing the no of electrical generating capacity in megawatts in 2014*

In [ ]:
plt.figure(figsize=(8,4))
sns.scatterplot(x='capacity_mw',y='generation_gwh_2015',data=df)

*In 2015 we can also notice here as electricity generation in gigawatt-hour increasing also increasing the no of electrical generating capacity in megawatts in 2015*

In [ ]:
plt.figure(figsize=(8,4))
sns.scatterplot(x='capacity_mw',y='generation_gwh_2016',data=df)

*In 2016 we can also notice here as electricity generation in gigawatt-hour increasing also increasing the no of electrical generating capacity in megawatts in 2016*

In [ ]:
plt.figure(figsize=(8,4))
sns.scatterplot(x='capacity_mw',y='generation_gwh_2017',data=df)

*In 2017 we can also notice here as electricity generation in gigawatt-hour increasing also increasing the no of electrical generating capacity in megawatts in 2017*

In [ ]:
plt.figure(figsize=(8,4))
sns.scatterplot(x='capacity_mw',y='latitude',data=df)

In [ ]:
plt.figure(figsize=(8,4))
sns.scatterplot(x='primary_fuel',y='generation_gwh_2013',data=df)

*-> We can here the coal is most used fuel in 2013 and Nuclear,Gas and hydro on average use and Solar,Wind Bomass nad oil was very less used*

In [ ]:
plt.figure(figsize=(8,4))
sns.scatterplot(x='primary_fuel',y='generation_gwh_2014',data=df)

In [ ]:
plt.figure(figsize=(8,4))
sns.scatterplot(x='primary_fuel',y='generation_gwh_2015',data=df)

In [ ]:
plt.figure(figsize=(8,4))
sns.scatterplot(x='primary_fuel',y='generation_gwh_2016',data=df)

In [ ]:
plt.figure(figsize=(8,4))
sns.scatterplot(x='primary_fuel',y='generation_gwh_2017',data=df)

*In all of the year 2013 to 2017 the coal is most used fuel in 2013 and Nuclear,Gas and hydro on average use and Solar,Wind Bomass nad oil was very less used*

In [ ]:
plt.figure(figsize=(8,4))
sns.scatterplot(x='primary_fuel',y='latitude',data=df)

In [ ]:
plt.figure(figsize=(8,4))
sns.scatterplot(x='capacity_mw',y='primary_fuel',data=df)

*Coal is on Max Electrical generating capacity in Megawatts*

# Using LabelEncoder for convering Categorical to Numerical

In [ ]:
data.info()

In [ ]:
data['primary_fuel'].unnique()

In [ ]:
data['geolocation_source'].unique()

In [ ]:
encoder = LabelEncoder()
data['primary_fuel'] = encoder.fit_transform(df['primary_fuel'])
data['geolocation_source'] = encoder.fit_transform(df['geolocation_source'])
data['source'] = encoder.fit_transform(df['source'])

In [ ]:
data['geolocation_source'].unique()

In [ ]:
data.info()

*By using zscore method for removal of OUTLIERS*

In [ ]:
from urllib.request import DataHandler
z=np.abs(zscore(data[['generation_gwh_2013','generation_gwh_2014','generation_gwh_2015','generation_gwh_2016','generation_gwh_2017','capacity_mw']]))
print(np.where(z>3))

In [ ]:
z = np.abs(zscore(df[['generation_gwh_2013','generation_gwh_2014','generation_gwh_2015','generation_gwh_2016','generation_gwh_2017','capacity_mw']]))
threshold = 3
df1 = df[(z<3).all(axis = 1)]

print ("Shape of the dataframe before removing outliers: ", df.shape)
print ("Shape of the dataframe after removing outliers: ", df1.shape)
print ("Percentage of data loss post outlier removal: ", (df.shape[0]-df1.shape[0])/df.shape[0]*100)

In [ ]:
 # reassigning the changed dataframe name to our original dataframe name
data=data1.copy()

In [ ]:
data.shape

*SKEWNESS HANDLING*

In [ ]:
plt.figure(figsize=(18,20))
for i in enumerate(df):
    plt.subplot(8,4,i[0]+1)
    sns.distplot(df[i[1]],color='g')

In [ ]:
data.skew()

skewness more than+/- 0.5 will we treated





object columns will no be treated

In [ ]:
df[['capacity_mw',"generation_gwh_2013",'generation_gwh_2014','generation_gwh_2015','generation_gwh_2016','generation_gwh_2017']]=power_transform(df[['capacity_mw',"generation_gwh_2013",'generation_gwh_2014','generation_gwh_2015','generation_gwh_2016','generation_gwh_2017']])
df.skew()

*First taking 'capacity_mw' as target creating Regression model*

*Spliting the Dataset*

In [ ]:
x=data.drop(['capacity_mw'],axis=1)
y=data['capacity_mw']

In [ ]:
x.shape

In [ ]:
y.shape

*Scaling X values*

In [ ]:
sc=MinMaxScaler()
x=sc.fit_transform(x)
pd.DataFrame(x).isnull().sum()

In [ ]:
pd.DataFrame(x).describe()

In [ ]:
pd.DataFrame(x).describe().T

*Data have been scaled*

# MODEL BUILDING



## Linear Regression

*CHECKING THE BEST RANDOM STATE FOR DATA SPLIT*

In [ ]:
LR=LinearRegression()

for i in range(0,100):
    x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=i)
    LR.fit(x_train,y_train)
    LR_predict_train=LR.predict(x_train)
    LR_predict_test=LR.predict(x_test)
    print(f'At random state {i}, The training accuracy is: {r2_score(y_train,LR_predict_train)}')
    print(f'At random state {i}, The test accuracy is: {r2_score(y_test,LR_predict_test)}')
    print('\n')

*random state=95 giving highest accuracy, choosing that as random state*

# SPLITTING THE DATA WITH RANDOM STATE =95

*Splitting the data into 73% training and 27% testing*

In [ ]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.27,random_state=95)

In [ ]:
x_train.shape

In [ ]:
x_test.shape

In [ ]:
y_train.shape

In [ ]:
y_test.shape

In [ ]:
LR=LinearRegression()
LR.fit(x_train,y_train)
print(LR.score(x_train,y_train))
LR_predict=LR.predict(x_test)

In [ ]:
print('MSE:',mean_squared_error(LR_predict,y_test))
print('MAE:',mean_absolute_error(LR_predict,y_test))
print('r2_score:',r2_score(LR_predict,y_test))

# Ridge

In [ ]:
R=Ridge()
R.fit(x_train,y_train)
print(R.score(x_train,y_train))
R_predict=R.predict(x_test)

In [ ]:
print('MSE:',mean_squared_error(R_predict,y_test))
print('MAE:',mean_absolute_error(R_predict,y_test))
print('r2_score:',r2_score(R_predict,y_test))

# **SVR**

# SVR(kernel='linear)

In [ ]:
svr=SVR(kernel='linear')
svr.fit(x_train,y_train)
print(svr.score(x_train,y_train))
svr_predict=svr.predict(x_test)

In [ ]:
print('MSE:',mean_squared_error(svr_predict,y_test))
print('MAE:',mean_absolute_error(svr_predict,y_test))
print('r2_score:',r2_score(svr_predict,y_test))

In [ ]:
svr_p=SVR(kernel='poly')
svr_p.fit(x_train,y_train)
print(svr_p.score(x_train,y_train))
svrpred_p=svr_p.predict(x_test)

In [ ]:
print('MSE:',mean_squared_error(svrpred_p,y_test))
print('MAE:',mean_absolute_error(svrpred_p,y_test))
print('r2_score:',r2_score(svrpred_p,y_test))

In [ ]:
svr_r=SVR(kernel='rbf')
svr_r.fit(x_train,y_train)
print(svr_r.score(x_train,y_train))
svrpred_r=svr_r.predict(x_test)

In [ ]:
print('MSE:',mean_squared_error(svrpred_r,y_test))
print('MAE:',mean_absolute_error(svrpred_r,y_test))
print('r2_score:',r2_score(svrpred_r,y_test))

# RandomForestRegressor

In [ ]:
RF=RandomForestRegressor()
RF.fit(x_train,y_train)
print(RF.score(x_train,y_train))
RF_PRED=RF.predict(x_test)

In [ ]:
print('MSE:',mean_squared_error(RF_PRED,y_test))
print('MAE:',mean_absolute_error(RF_PRED,y_test))
print('r2_score:',r2_score(RF_PRED,y_test))

# DecisionTreeRegressor

In [ ]:
DTR=DecisionTreeRegressor()
DTR.fit(x_train,y_train)
print(DTR.score(x_train,y_train))
DTR_PRED=DTR.predict(x_test)

In [ ]:
print('MSE:',mean_squared_error(DTR_PRED,y_test))
print('MAE:',mean_absolute_error(DTR_PRED,y_test))
print('r2_score:',r2_score(DTR_PRED,y_test))

# GradientBoostingRegressor

In [ ]:
GBR=GradientBoostingRegressor()
GBR.fit(x_train,y_train)
print(GBR.score(x_train,y_train))
GBR_PRED=GBR.predict(x_test)

In [ ]:
print('MSE:',mean_squared_error(GBR_PRED,y_test))
print('MAE:',mean_absolute_error(GBR_PRED,y_test))
print('r2_score:',r2_score(GBR_PRED,y_test))

# Cross Validation

In [ ]:
np.random.seed(10)
def rmse_cv(model, x,y):
    rmse =- (cross_val_score(model, x,y, scoring='neg_mean_squared_error', cv=10))
    return(rmse)


models = [LinearRegression(),
             Ridge(),
             SVR(kernel='linear'),
             SVR(kernel='poly'),
             SVR(kernel='rbf'),
             RandomForestRegressor(),
             DecisionTreeRegressor(),
             GradientBoostingRegressor(),]



names = ['LR','R','svr','svr_p','svr_r','RF','DTR','GBR']

for model,name in zip(models,names):
    score = rmse_cv(model,x,y)
    print("{}    : {:.6f}, {:4f}".format(name,score.mean(),score.std()))

# HYPER PARAMETER TUNING:

GridSearchCV

# With GradientBoostingRegressor


In [ ]:
GF=GradientBoostingRegressor()
param={
     'criterion':['mse','mae'],
     'n_estimators':[100, 200],
    }

In [ ]:
GF_grid=GridSearchCV(RandomForestRegressor(),param,cv=10,scoring='accuracy',n_jobs=-1,verbose=2)

In [ ]:
GF_grid.fit(x_train,y_train)
GF_grid_pred=GF_grid.best_estimator_.predict(x_test)

In [ ]:
GF_grid.best_params_

In [ ]:
print('MSE:',mean_squared_error(GF_grid_pred,y_test))
print('MAE:',mean_absolute_error(GF_grid_pred,y_test))
print('r2_score:',r2_score(GF_grid_pred,y_test))

In [ ]:
sns.distplot(GF_grid_pred-y_test)

In [ ]:
plt.scatter(GF_grid_pred,y_test)
plt.plot(y_test,y_test,linewidth=2,color='Red')

# Choosing GradientBoostingRegressor as the best model

# TAKING 'primary_fuel' AS TARGET , CREATING CLASSIFICATION MODEL Dividing data into X and Y

In [ ]:
x=df.drop(['primary_fuel'],axis=1)
y=df['primary_fuel']

In [ ]:
x.shape

In [ ]:
y.shape

# Spliting Train and Test data

In [ ]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=.27,random_state=42)

In [ ]:
x_train.shape

In [ ]:
x_test.shape

In [ ]:
y_train.shape

In [ ]:
y_test.shape

# **MODEL BUILDING**

In [ ]:
#Importing necessary libraries

from sklearn.linear_model import LogisticRegression

from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score

# LogisticRegression

In [ ]:
logreg=LogisticRegression()
logreg_score=cross_val_score(logreg,x,y,cv=5,scoring='accuracy')
print("cross validation score:",np.mean(logreg_score))

In [ ]:
logreg.fit(x_train,y_train)
LR_predicted=logreg.predict(x_test)

print(accuracy_score(y_test,LR_predicted))
print(confusion_matrix(y_test,LR_predicted))
print(classification_report(y_test,LR_predicted))
print("Training accuracy::",logreg.score(x_train,y_train))
print("Test accuracy::",logreg.score(x_test,y_test))

# DECISION TREE CLASSIFIER

In [ ]:
dtc=DecisionTreeClassifier()
dtc_score=cross_val_score(dtc,x,y,cv=5,scoring='accuracy')
print("cross validation score for svm:",np.mean(dtc_score))

In [ ]:
dtc.fit(x_train,y_train)
predicted_dtc=dtc.predict(x_test)
print(accuracy_score(y_test,predicted_dtc))
print(confusion_matrix(y_test,predicted_dtc))
print(classification_report(y_test,predicted_dtc))
print("Training accuracy::",dtc.score(x_train,y_train))
print("Test accuracy::",dtc.score(x_test,y_test))

# KNeighborsClassifier

In [ ]:
knn=KNeighborsClassifier()
knn_score=cross_val_score(knn,x,y,cv=5,scoring='accuracy')
print("cross validation score for knn:",np.mean(knn_score))

In [ ]:
knn.fit(x_train,y_train)
predicted_knn=knn.predict(x_test)
print(accuracy_score(y_test,predicted_knn))
print(confusion_matrix(y_test,predicted_knn))
print(classification_report(y_test,predicted_knn))
print("Training accuracy::",knn.score(x_train,y_train))
print("Test accuracy::",knn.score(x_test,y_test))

# RandomForestClassifier

In [ ]:
rfc=RandomForestClassifier()
rfc_score=cross_val_score(rfc,x,y,cv=5,scoring='accuracy')
print("cross validation score for rfc:",np.mean(rfc_score))

In [ ]:
rfc.fit(x_train,y_train)
predicted_rfc=rfc.predict(x_test)
print(accuracy_score(y_test,predicted_rfc))
print(confusion_matrix(y_test,predicted_rfc))
print(classification_report(y_test,predicted_rfc))
print("Training accuracy::",rfc.score(x_train,y_train))
print("Test accuracy::",rfc.score(x_test,y_test))

# ENSEMBLE TECHNIQUES


# 1. ADA BOOST CLASSIFIER

In [ ]:
adb=AdaBoostClassifier()
adb_score=cross_val_score(adb,x,y,cv=10,scoring='accuracy')
print("cross validation score for Ada boost:",np.mean(adb_score))

In [ ]:
adb.fit(x_train,y_train)
predicted_adb=adb.predict(x_test)
print(accuracy_score(y_test,predicted_adb))
print(confusion_matrix(y_test,predicted_adb))
print(classification_report(y_test,predicted_adb))
print("Training accuracy::",adb.score(x_train,y_train))
print("Test accuracy::",adb.score(x_test,y_test))

# 2. BAGGING CLASSIFIER

In [ ]:
bgc=BaggingClassifier()
bgc_score=cross_val_score(bgc,x,y,cv=10,scoring='accuracy')
print("cross validation score for BAGGING Classifier:",np.mean(bgc_score))

In [ ]:
bgc.fit(x_train,y_train)
predicted_bgc=bgc.predict(x_test)
print(accuracy_score(y_test,predicted_bgc))
print(confusion_matrix(y_test,predicted_bgc))
print(classification_report(y_test,predicted_bgc))
print("Training accuracy::",bgc.score(x_train,y_train))
print("Test accuracy::",bgc.score(x_test,y_test))

# 3. Gradient Boosting classifier



In [ ]:
grbc=GradientBoostingClassifier()
grbc_score=cross_val_score(grbc,x,y,cv=10,scoring='accuracy')
print("cross validation score for Gradient Boosting Classifier:",np.mean(grbc_score))

In [ ]:
grbc.fit(x_train,y_train)
predicted_grbc=grbc.predict(x_test)
print(accuracy_score(y_test,predicted_grbc))
print(confusion_matrix(y_test,predicted_grbc))
print(classification_report(y_test,predicted_grbc))
print("Training accuracy::",grbc.score(x_train,y_train))
print("Test accuracy::",grbc.score(x_test,y_test))

# HYPER PARAMETER TUNING:

In [ ]:
rfc=RandomForestClassifier()
param_grid={
    'criterion' : ["gini", "entropy"],
    'n_estimators':[200,50],
    'min_samples_split':[2,3],
     'random_state':[10]
}

In [ ]:
rfc_grid=GridSearchCV(RandomForestClassifier(),param_grid,cv=4,scoring='accuracy',n_jobs=-1,verbose=2)

In [ ]:
rfc_grid.fit(x_train,y_train)
rfc_pred=rfc_grid.best_estimator_.predict(x_test)
print("Accuracy after parameter tuning::",accuracy_score(y_test,rfc_pred))

In [ ]:
rfc_grid.best_params_

# Model training with best parameters

In [ ]:
best_param={
    'criterion' : ['gini'],
    'n_estimators':[200],
    'min_samples_split':[2],
     'random_state':[10]    }

In [ ]:
best_rfc_grid=GridSearchCV(RandomForestClassifier(),best_param,cv=4,scoring='accuracy',n_jobs=-1,verbose=2)

In [ ]:
best_rfc_grid.fit(x_train,y_train)
best_rfc_pred=best_rfc_grid.best_estimator_.predict(x_test)
print("Accuracy after parameter tuning::",accuracy_score(y_test,best_rfc_pred))

# Report of RandomForestClassifier

In [ ]:
print("CLASSIFICATION REPORT::\n",classification_report(y_test,best_rfc_pred))

In [ ]:
clsf_repo = classification_report(y_test,best_rfc_pred,output_dict=True)
sns.heatmap(pd.DataFrame(clsf_repo).T, annot=True)

In [ ]:
y_prob = best_rfc_grid.fit(x_train,y_train).predict_proba(x_test)

macro_roc_auc_ovo = roc_auc_score(y_test, y_prob, multi_class="ovo", average="macro")
weighted_roc_auc_ovo = roc_auc_score(y_test, y_prob, multi_class="ovo", average="weighted")
macro_roc_auc_ovr = roc_auc_score(y_test, y_prob, multi_class="ovr", average="macro")
weighted_roc_auc_ovr = roc_auc_score(y_test, y_prob, multi_class="ovr", average="weighted")
print("One-vs-One ROC AUC scores:\n{:.6f} (macro),\n{:.6f} "
      "(weighted by prevalence)"
      .format(macro_roc_auc_ovo, weighted_roc_auc_ovo))
print("="*40)
print("One-vs-Rest ROC AUC scores:\n{:.6f} (macro),\n{:.6f} "
      "(weighted by prevalence)"
      .format(macro_roc_auc_ovr, weighted_roc_auc_ovr))

# Saving Model:

In [ ]:
import joblib
joblib.dump(best_rfc_grid.best_estimator_,'Global Power Plant Classification.obj')